Client/Frontend/Browser/SearchLoader.swift

```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import Storage
+
+private let URLBeforePathRegex = NSRegularExpression(pattern: "^https?://([^/]+/)", options: nil, error: nil)!
+
+// TODO: Swift current requires that classes extending generic classes must also be generic.
+// This is a workaround until that requirement is fixed.
+typealias SearchLoader = _SearchLoader<AnyObject, AnyObject>
+
+/**
+ * Shared data source for the SearchViewController and the URLBar domain completion.
+ * Since both of these use the same SQL query, we can perform the query once and dispatch the results.
+ */
+class _SearchLoader<UnusedA, UnusedB>: Loader<Cursor, SearchViewController> {
+    private let history: History
+    private let urlBar: URLBarView
```

Comment:
I don't think it's that strange -- this is a loader specific to the `SearchViewController` and `URLBarView`, so it needs to know those views to interact with them directly. This could have all just gone into BVC, but I figured you'd want them pulled out to reduce clutter as you mentioned before :)

I think it's important that the reverse dependencies not be coupled, though (that is, `SearchViewController` and `URLBarView` shouldn't know anything about their data sources).